File: gateway/internal/bus/bus.go
// Package bus membungkus Redis Streams (message.inbound/outbound/status,
// consumer group + ack) dan Pub/Sub realtime. Lihat docs/prd/01, 09.
package bus

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/chatcepat/gateway/internal/contracts"
	"github.com/redis/go-redis/v9"
)

func msDur(ms int) time.Duration { return time.Duration(ms) * time.Millisecond }

const (
	StreamInbound  = "message.inbound"
	StreamOutbound = "message.outbound"
	StreamStatus   = "message.status"
	GroupGateway   = "gateway"
	realtimePrefix = "realtime."
)

type Bus struct {
	rdb *redis.Client
}

func New(redisURL string) (*Bus, error) {
	opt, err := redis.ParseURL(redisURL)
	if err != nil {
		return nil, fmt.Errorf("parse redis url: %w", err)
	}
	return &Bus{rdb: redis.NewClient(opt)}, nil
}

func (b *Bus) Ping(ctx context.Context) error { return b.rdb.Ping(ctx).Err() }

func (b *Bus) Close() error { return b.rdb.Close() }

// EnsureGroup membuat consumer group (idempoten — abaikan BUSYGROUP).
func (b *Bus) EnsureGroup(ctx context.Context, stream, group string) error {
	err := b.rdb.XGroupCreateMkStream(ctx, stream, group, "0").Err()
	if err != nil && err.Error() != "BUSYGROUP Consumer Group name already exists" {
		return err
	}
	return nil
}

func (b *Bus) publish(ctx context.Context, stream string, v any) (string, error) {
	data, err := json.Marshal(v)
	if err != nil {
		return "", err
	}
	return b.rdb.XAdd(ctx, &redis.XAddArgs{
		Stream: stream,
		Values: map[string]any{"data": string(data)},
	}).Result()
}

// PublishInbound: pesan masuk ternormalisasi → engine.
func (b *Bus) PublishInbound(ctx context.Context, msg *contracts.InboundMessage) (string, error) {
	return b.publish(ctx, StreamInbound, msg)
}

// PublishStatus: status pengiriman → engine yang persist.
func (b *Bus) PublishStatus(ctx context.Context, st *contracts.MessageStatus) (string, error) {
	return b.publish(ctx, StreamStatus, st)
}

// PublishOutbound: perintah kirim → stream message.outbound. Normalnya dipublish engine;
// disediakan untuk test / tooling.
func (b *Bus) PublishOutbound(ctx context.Context, cmd *contracts.OutboundCommand) (string, error) {
	return b.publish(ctx, StreamOutbound, cmd)
}

// StatusTail mengembalikan N entri terakhir stream message.status (untuk test/inspeksi).
func (b *Bus) StatusTail(ctx context.Context, n int64) ([]contracts.MessageStatus, error) {
	res, err := b.rdb.XRevRangeN(ctx, StreamStatus, "+", "-", n).Result()
	if err != nil {
		return nil, err
	}
	out := make([]contracts.MessageStatus, 0, len(res))
	for _, m := range res {
		raw, _ := m.Values["data"].(string)
		var st contracts.MessageStatus
		if json.Unmarshal([]byte(raw), &st) == nil {
			out = append(out, st)
		}
	}
	return out, nil
}

// OutMsg = satu perintah kirim dari stream message.outbound.
type OutMsg struct {
	ID  string
	Cmd contracts.OutboundCommand
}

// ConsumeOutbound membaca perintah kirim untuk consumer (group gateway).
func (b *Bus) ConsumeOutbound(ctx context.Context, consumer string, count int64, blockMs int) ([]OutMsg, error) {
	res, err := b.rdb.XReadGroup(ctx, &redis.XReadGroupArgs{
		Group:    GroupGateway,
		Consumer: consumer,
		Streams:  []string{StreamOutbound, ">"},
		Count:    count,
		Block:    msDur(blockMs),
	}).Result()
	if err != nil {
		if err == redis.Nil {
			return nil, nil
		}
		return nil, err
	}
	var out []OutMsg
	for _, s := range res {
		for _, m := range s.Messages {
			raw, _ := m.Values["data"].(string)
			var cmd contracts.OutboundCommand
			if err := json.Unmarshal([]byte(raw), &cmd); err != nil {
				// skip pesan rusak tapi ack supaya tak loop (ditangani caller via ID)
				out = append(out, OutMsg{ID: m.ID})
				continue
			}
			out = append(out, OutMsg{ID: m.ID, Cmd: cmd})
		}
	}
	return out, nil
}

func (b *Bus) AckOutbound(ctx context.Context, id string) error {
	return b.rdb.XAck(ctx, StreamOutbound, GroupGateway, id).Err()
}

// PublishRealtime: fire-and-forget ke dashboard per tenant.
func (b *Bus) PublishRealtime(ctx context.Context, tenantID string, payload []byte) error {
	return b.rdb.Publish(ctx, realtimePrefix+tenantID, payload).Err()
}

// Subscribe ke realtime.<tenant> (dipakai WS hub).
func (b *Bus) Subscribe(ctx context.Context, tenantID string) *redis.PubSub {
	return b.rdb.Subscribe(ctx, realtimePrefix+tenantID)
}

File: engine/app/bus.py
"""Redis Streams (consumer group + ack) + Pub/Sub realtime.

Pesan stream disimpan di field `data` berisi JSON. Lihat docs/prd/01, 09.
"""

from __future__ import annotations

import redis.asyncio as redis
from redis.exceptions import ResponseError

from .config import (
    CONSUMER_GROUP,
    REALTIME_PREFIX,
    REDIS_URL,
    STREAM_OUTBOUND,
    STREAM_STATUS,
)

_client: redis.Redis | None = None


def client() -> redis.Redis:
    global _client
    if _client is None:
        _client = redis.from_url(REDIS_URL, decode_responses=True)
    return _client


async def ensure_group(stream: str, group: str = CONSUMER_GROUP) -> None:
    """Buat consumer group (idempoten — abaikan BUSYGROUP)."""
    try:
        await client().xgroup_create(stream, group, id="0", mkstream=True)
    except ResponseError as e:
        if "BUSYGROUP" not in str(e):
            raise


async def read_group(
    stream: str,
    consumer: str,
    group: str = CONSUMER_GROUP,
    count: int = 10,
    block_ms: int = 5000,
) -> list[tuple[str, dict[str, str]]]:
    """Baca pesan baru untuk consumer. Return list (message_id, fields)."""
    resp = await client().xreadgroup(
        group, consumer, {stream: ">"}, count=count, block=block_ms
    )
    out: list[tuple[str, dict[str, str]]] = []
    for _stream, messages in resp or []:
        for msg_id, fields in messages:
            out.append((msg_id, fields))
    return out


async def ack(stream: str, msg_id: str, group: str = CONSUMER_GROUP) -> None:
    await client().xack(stream, group, msg_id)


async def publish(stream: str, payload: str) -> str:
    return await client().xadd(stream, {"data": payload})


async def publish_outbound(payload_json: str) -> str:
    return await publish(STREAM_OUTBOUND, payload_json)


async def publish_status(payload_json: str) -> str:
    return await publish(STREAM_STATUS, payload_json)


async def publish_realtime(tenant_id: str, payload_json: str) -> int:
    """Fire-and-forget realtime ke channel per tenant."""
    return await client().publish(f"{REALTIME_PREFIX}{tenant_id}", payload_json)


async def close() -> None:
    global _client
    if _client is not None:
        await _client.aclose()
        _client = None
